### PR TITLE
Migrate lists/extract to pylibcudf

### PR DESCRIPTION
**Pull request by Matt711**
## Description

Apart of #15162 
## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/_lib/pylibcudf/lists.pyx [hunk 1]
```diff
@@ -94,3 +100,29 @@ cpdef Column concatenate_list_elements(Column input, bool dropna):
...
+    Parameters
+    ----------
+    input : Column
+        The input column.
+    index : Union[Column, size_type]
+        The selection index or indices.
+    Returns
+    -------
+    Column
+        A new Column of bools
```
Comment (wence-) [inline python/cudf/cudf/_lib/pylibcudf/lists.pyx:116]:
```suggestion
    """Create a column of extracted list elements.
    Parameters
    ----------
    input : Column
        The input column.
    index : Union[Column, size_type]
        The selection index or indices.
    Returns
    -------
    Column
        A new Column of bools
```
```suggestion
    """Create a column of extracted list elements.

    
    Parameters
    ----------
    input : Column
        The input column.
    index : Union[Column, size_type]
        The selection index or indices.
    
    Returns
    -------
    Column
        A new Column with elements extracted.
```

File: python/cudf/cudf/_lib/pylibcudf/lists.pyx [hunk 2]
```diff
@@ -94,3 +100,29 @@ cpdef Column concatenate_list_elements(Column input, bool dropna):
...
+    cdef unique_ptr[column] c_result
+    cdef shared_ptr[lists_column_view] list_view = (
+        make_shared[lists_column_view](input.view())
+    )
+    if ColumnOrSizeType is Column:
+        with nogil:
+            c_result = move(cpp_extract_list_element(list_view.get()[0], index.view()))
+    else:
+        with nogil:
+            c_result = move(cpp_extract_list_element(list_view.get()[0], index))
```
Comment (wence-) [inline python/cudf/cudf/_lib/pylibcudf/lists.pyx:127]: *issue*: I think unfortunately, we need to handle the case that the argument is `object` (see here https://github.com/rapidsai/cudf/pull/15257#discussion_r1610591457)

@vyasr I think we should put something in the dev guide.

Comment (vyasr) [reply python/cudf/cudf/_lib/pylibcudf/lists.pyx:127]: Yes, we should. I will also try and set aside a bit of time to investigate fixing this upstream. It's quite annoying how far we're having to propagate this workaround.

Comment (Matt711) [reply python/cudf/cudf/_lib/pylibcudf/lists.pyx]: [`size_type`](https://docs.rapids.ai/api/libcudf/stable/group__utility__types#ga0aff16b2db009d92300befaca11827e9) is not a `cpdef enum class`. So we wouldn't need to check for `object`, right?

File: python/cudf/cudf/_lib/pylibcudf/lists.pyx [hunk 3]
```diff
@@ -206,3 +209,29 @@ cpdef Column index_of(Column input, ColumnOrScalar search_key, bool find_first_o
...
+        ----------
+        input : Column
+            The input column.
+        index : Union[Column, size_type]
+            The selection index or indices.
+
+        Returns
+        -------
+        Column
+            A new Column with elements extracted.
```
Comment (wence-) [inline python/cudf/cudf/_lib/pylibcudf/lists.pyx:227]:
```suggestion
    Parameters
    ----------
    input : Column
        The input column.
    index : Union[Column, size_type]
        The selection index or indices.

    Returns
    -------
    Column
        A new Column with elements extracted.
```

File: python/cudf/cudf/pylibcudf_tests/test_lists.py [hunk 4]
```diff
@@ -134,3 +134,25 @@ def test_index_of_list_column(test_data, column):
...
+            -1,
+            pa.array([1, 2, 5, 7]),
+        ),
+    ],
+)
+def test_extract_list_element(test_data, index, expect):
+    arr = pa.array(test_data[0][0])
+    plc_column = plc.interop.from_arrow(arr)
+
+    res = plc.lists.extract_list_element(plc_column, index)
```
[the remark this answers is not in the document]
Comment (Matt711) [reply python/cudf/cudf/pylibcudf_tests/test_lists.py]: Thanks, it looks that arrow function is not fully implemented.

**Pull request merged**